runelite-client/src/main/java/net/runelite/client/plugins/inferno/InfernoPlugin.java

```diff
@@ -98,49 +98,140 @@
 	private EventBus eventBus;
 
 	@Getter(AccessLevel.PACKAGE)
-	private int currentWave = -1;
+	private InfernoConfig.FontStyle fontStyle = InfernoConfig.FontStyle.BOLD;
+	@Getter(AccessLevel.PACKAGE)
+	private int textSize = 32;
+
+	private WorldPoint lastLocation = new WorldPoint(0, 0, 0);
 
 	@Getter(AccessLevel.PACKAGE)
-	private final Map<NPC, InfernoNPC> monsters = new HashMap<>();
+	private int currentWaveNumber;
 
 	@Getter(AccessLevel.PACKAGE)
-	private final Map<Integer, ArrayList<InfernoNPC>> monsterCurrentAttackMap = new HashMap<>(6);
+	private final List<InfernoNPC> infernoNpcs = new ArrayList<>();
 
 	@Getter(AccessLevel.PACKAGE)
-	private final List<NPC> nibblers = new ArrayList<>();
+	private final HashMap<Integer, HashMap<InfernoNPC.Attack, Integer>> upcomingAttacks = new HashMap<>();
```

Comment:
`HashMap<Integer, Pair<InfernoNPC.Attack, Integer>>` would not work as multiple types of attacks (melee, range, magic) need to be stored within one tick (the key of the first map). I can look into your second suggestion but this would need a lot of refactoring.

Reply:
whoops, I noticed that after I typed out the example but only changed it at the array.

The biggest problem I have with the way you currently have it is that you're clearing the root map every single game tick, while you can be sure the ones with a delay of >= 1 will still be there, just decremented (or index incremented :P).

We could also take it even further, by having a `int[10][4]`, and adding index in the second array to the attackstyle enum. Shouldn't priority be cumulative by the way?

Reply:
Technically yes, but there's a lot of exceptions such as an NPC dying or simply taking cover behind one of the pillars. In these cases the ticks couldn't just be decremented so additional checks would be necessary.
Because these scenarios happen a lot, it's hard to tell if it the slight performance boost is worth the clutter of code that comes with the additional checks. Not to mention the blob, which is incredibly frustrating as it is and will most likely need to be completely redone in a new system.

Reply:
This should also use an interface instead of an implementation.

`private final Map<Integer, Map<InfernoNPC.Attack, Integer>> upcomingAttacks = new HashMap<>();`

Instead of

`private final HashMap<Integer, HashMap<InfernoNPC.Attack, Integer>> upcomingAttacks = new HashMap<>();`